src/gui/guiScene.cpp

```diff
@@ -0,0 +1,262 @@
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "guiScene.h"
+
+#include <SViewFrustum.h>
+#include <IAnimatedMeshSceneNode.h>
+#include <ILightSceneNode.h>
+#include "porting.h"
+
+GUIScene::GUIScene(gui::IGUIEnvironment *env, scene::ISceneManager *smgr,
+		   gui::IGUIElement *parent, core::recti rect, s32 id)
+	: IGUIElement(gui::EGUIET_ELEMENT, env, parent, id, rect)
+{
+	m_driver = env->getVideoDriver();
+	m_smgr = smgr->createNewSceneManager(false);
+
+	m_cam = m_smgr->addCameraSceneNode(0, v3f(0.f, 0.f, -100.f), v3f(0.f));
+	m_cam->setFOV(30.f * core::DEGTORAD);
+
+	scene::ILightSceneNode *light = m_smgr->addLightSceneNode(m_cam);
+	light->setRadius(1000.f);
+
+	m_smgr->getParameters()->setAttribute(scene::ALLOW_ZWRITE_ON_TRANSPARENT, true);
+}
+
+GUIScene::~GUIScene()
+{
+	setMesh(nullptr);
+
+	m_smgr->drop();
+}
+
+scene::IAnimatedMeshSceneNode *GUIScene::setMesh(scene::IAnimatedMesh *mesh)
+{
+	if (m_mesh) {
+		m_mesh->remove();
+		m_mesh = nullptr;
+	}
+
+	if (!mesh)
+		return nullptr;
+
+	m_mesh = m_smgr->addAnimatedMeshSceneNode(mesh);
+	m_mesh->setPosition(-m_mesh->getBoundingBox().getCenter());
+	m_mesh->animateJoints();
+	return m_mesh;
+}
+
+void GUIScene::setTexture(u32 idx, video::ITexture *texture)
+{
+	video::SMaterial &material = m_mesh->getMaterial(idx);
+	material.MaterialType = video::EMT_TRANSPARENT_ALPHA_CHANNEL;
+	material.MaterialTypeParam = 0.5f;
+	material.TextureLayer[0].Texture = texture;
+	material.setFlag(video::EMF_LIGHTING, false);
+	material.setFlag(video::EMF_FOG_ENABLE, true);
+	material.setFlag(video::EMF_BILINEAR_FILTER, false);
+	material.setFlag(video::EMF_BACK_FACE_CULLING, false);
+}
+
+void GUIScene::draw()
+{
+	// Control rotation speed based on time
+	u64 new_time = porting::getTimeMs();
+	u64 dtime_ms = 0;
+	if (m_last_time != 0)
+		dtime_ms = porting::getDeltaMs(m_last_time, new_time);
+	m_last_time = new_time;
+
+	core::rect<s32> oldViewPort = m_driver->getViewPort();
+	m_driver->setViewPort(getAbsoluteClippingRect());
+	core::recti borderRect = Environment->getRootGUIElement()->getAbsoluteClippingRect();
+
+	if (m_bgcolor != 0) {
+		Environment->getSkin()->draw3DSunkenPane(
+			this, m_bgcolor, false, true, borderRect, 0);
+	}
+
+	core::dimension2d<s32> size = getAbsoluteClippingRect().getSize();
+	m_smgr->getActiveCamera()->setAspectRatio((f32)size.Width / (f32)size.Height);
+
+	if (!m_target) {
+		updateCamera(m_smgr->addEmptySceneNode());
+		rotateCamera(v3f(0.f));
+		m_cam->bindTargetAndRotation(true);
+	}
+
+	cameraLoop();
+
+	// Continuous rotation
+	if (m_inf_rot)
+		rotateCamera(v3f(0.f, -0.03f * (float)dtime_ms, 0.f));
+
+	m_smgr->drawAll();
+
+	if (m_initial_rotation && m_mesh) {
+		rotateCamera(v3f(m_custom_rot.X, m_custom_rot.Y, 0.f));
+		calcOptimalDistance();
+
+		m_initial_rotation = false;
+	}
+
+	m_driver->setViewPort(oldViewPort);
+}
+
+bool GUIScene::OnEvent(const SEvent &event)
+{
+	if (m_mouse_ctrl) {
+		switch(event.EventType) {
```

Comment:
Why not just an `if (event.EventType == EET_MOUSE_INPUT_EVENT)`?  There's only one branch in the switch.

If the switch is kept, then `switch(...)` --> `switch (...)`.